store/src/java/com/zimbra/cs/service/PreviewServlet.java

```diff
@@ -0,0 +1,576 @@
+
+  /**
+   * This method is used to send error response for {@link HttpServletRequest}
+   *
+   * @param resp the {@link HttpServletResponse} object
+   * @param errCode error code for {@link HttpServletResponse}
+   * @param reason message string for {@link HttpServletResponse}
+   */
+  void respondWithError(HttpServletResponse resp, int errCode, String reason) {
+    resp.setContentType("text/html; charset=UTF-8");
+    try {
+      resp.sendError(errCode, reason);
+    } catch (IOException e) {
+      LOG.error(e.getMessage(), e);
+    }
+  }
+
+  @Override
+  protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+      throws IOException, ServletException {
+    ZimbraLog.clearContext();
+    addRemoteIpToLoggingContext(req);
+
+    final AuthToken authToken = getAuthTokenFromCookie(req, resp);
+    checkAuthTokenFromCookieOrRespondWithError(authToken, req, resp);
+
+    final Pattern requiredQueryParametersPattern =
+        Pattern.compile(SERVLET_PATH + "/([a-zA-Z]+)/([0-9]+)/([0-9]+)");
+    final Matcher requiredQueryParametersMatcher =
+        requiredQueryParametersPattern.matcher(getUrlWithQueryParams(req));
+
+    // check url for the presence of required parameters and query string
+    // send error otherwise
+    if (req.getQueryString() == null
+        || (!requiredQueryParametersMatcher.find()
+            || requiredQueryParametersMatcher.groupCount() != 3)) {
+      respondWithError(resp, HttpServletResponse.SC_BAD_REQUEST, "Missing required parameters");
+    } else {
+      final int messageId = Integer.parseInt(requiredQueryParametersMatcher.group(2));
+      final String partNo = requiredQueryParametersMatcher.group(3);
+
+      // get attachment
+      final Try<MimePart> attachmentMimePart = getAttachment(authToken, messageId, partNo);
+      if (attachmentMimePart.isFailure()) {
+        respondWithError(
+            resp,
+            HttpServletResponse.SC_INTERNAL_SERVER_ERROR,
+            attachmentMimePart.failed().get().getMessage());
+      }
+
+      // send preview response or commit preview api error as response
+      Try.of(() -> getAttachmentPreview(getUrlWithQueryParams(req), attachmentMimePart.get()))
+          .onSuccess(previewOfAttachment -> respondWithSuccess(resp, previewOfAttachment.get()))
+          .onFailure(
+              ex -> {
+                if (ex instanceof BadRequest) {
+                  respondWithError(resp, HttpServletResponse.SC_BAD_REQUEST, ex.getMessage());
+                } else if (ex instanceof ItemNotFound) {
+                  respondWithError(resp, HttpServletResponse.SC_NOT_FOUND, ex.getMessage());
+                } else if (ex instanceof ValidationError) {
+                  respondWithError(resp, HttpServletResponse.SC_UNAUTHORIZED, ex.getMessage());
+                } else {
+                  respondWithError(
+                      resp, HttpServletResponse.SC_INTERNAL_SERVER_ERROR, ex.getMessage());
+                }
+              });
```

Comment:
This can still be refined in:

```suggestion
        Try.of(() -> getAttachmentPreview(getUrlWithQueryParams(req), attachmentMimePart.get()))
                .onSuccess(previewOfAttachment -> respondWithSuccess(resp, previewOfAttachment.get()))
                .onFailure(BadRequest.class, ex -> respondWithError(resp, HttpServletResponse.SC_BAD_REQUEST, ex.getMessage()))
                .onFailure(ItemNotFound.class, ex -> respondWithError(resp, HttpServletResponse.SC_NOT_FOUND, ex.getMessage()))
                .onFailure(ValidationError.class, ex -> respondWithError(resp, HttpServletResponse.SC_UNAUTHORIZED, ex.getMessage())
                .onFailure(ex -> respondWithError(resp, HttpServletResponse.SC_INTERNAL_SERVER_ERROR, ex.getMessage());
```

This way it's easier to handle new errors in case it's needed and it's clear what's happening. Plus, you are not doing type checking, that it is known for being slow.

Reply:
I tested this when you suggested it in the previous comment. Unfortunately, this results in undesirable behavior. I am getting two error codes :smiling_face_with_tear:

Reply:
Can you print them here? It seems strange, I tested this code on my machine and they worked :sweat_smile: